src/electron/preload.ts

```diff
@@ -0,0 +1,81 @@
+// Copyright 2018 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Starting from electron 12, contextIsolation is turned on by default.
+// So we are not able to use node APIs or electron APIs in renderer scripts
+// any more. We have to inject key features into the global window object
+// in preload: https://www.electronjs.org/docs/latest/tutorial/tutorial-preload.
+
+import {contextBridge, ipcRenderer} from 'electron';
+import * as os from 'os';
+import {ShadowsocksConfig} from '../www/app/config';
+import {ErrorCode} from '../www/model/errors';
+
+contextBridge.exposeInMainWorld('os', {
+  platform: os.platform(),
+});
+
+contextBridge.exposeInMainWorld('ipc', {
```

Comment:
Can we throw in the preload script? Perhaps we could do the throw here, so we can preserve the use of exceptions.

Reply:
We cannot throw errors in preload, more information here: https://github.com/electron/electron/issues/25596#issuecomment-700281408

Reply:
But I think one thing we can do is wrapping all IPC calling in a function called `invokeIpc(func, ...args)` in util. So we can wrap this error marshaling logic in that function.

Reply:
BTW, I think you can throw errors, but they are limited to a few types: https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm

Reply:
Never mind, I wrapped the exception logic in `OutlineIpcClient`, so the caller does not need to know anything about `ErrorCode` at all.